FHEM/00_SIGNALduino.pm

```diff
@@ -3651,17 +3652,31 @@ sub SIGNALduino_OSV2()
 		#$dmsg=$osv2hex;
 		return (1,$osv2hex);
 	}
-	elsif ($bitData =~ m/^.?(1){16,24}0101/)  {  # Valid OSV3 detected!	
-		$preamble_pos = index($bitData, '0101', 16);
-		$message_end = length($bitData);
-		$message_length = $message_end - ($preamble_pos+4);
-		SIGNALduino_Log3 $name, 4, "$name: OSV3 protocol detected: preamble_pos = $preamble_pos, message_length = $message_length";
+	elsif (($preamble_pos = index($bitData, '1111111111110101')) >= 0)  {  # Preamble 12 x 1, Valid OSV3 detected!	
```

Comment:
warum hast Du es nicht bei der Suche via regex gelassen?

Reply:
index ist die schnellste Variante um was zu suchen. Die regex erschien mir unnötig aufwendig.
Vorher waren zwei suchen (regex + index) hintereinander drin.

Reply:
Die Regex hat auf die Anzahl der "1" ein min/max gelegt. Die Regex hätte bei 25 x1 nicht wahr ergeben. Die Suche via Index kennt keine mindest und Maximallänge und liefert auch wahr zurück, wenn es mehr als die maximale Anzahl ist.
Für den PR hätte man doch nur die Mindestlänge von 16 auf 12 herabsetzen müssen oder?

Die 2. Suche hätte man sich in der Tat schenken können, da stimme ich dir zu. Die war nicht wirklich nötig. Das Ergebnis steht schon in `$+[0]`

Reply:
>Die Regex hat auf die Anzahl der "1" ein min/max gelegt. Die Regex hätte bei 25 x1 nicht wahr ergeben. Die Suche via Index kennt keine mindest und Maximallänge und liefert auch wahr zurück, wenn es mehr als die maximale Anzahl ist.

Mir ist nicht klar welchen Fall Du mit der Abfrage der maximal Länge abdecken willst.
Es müsste doch die folgende Abfrage reichen:
```
elsif ($bitData =~ m/(1){12}0101/)  {  # Valid OSV3 detected!
$preamble_pos = $+[0];
```
Ich möchte gerne die folgenden Fälle abdecken:
erste Nachricht mit preamble, zweite Nachricht zu kurz  
`MC;LL=-975;LH=984;SL=-487;SH=482;D=000283AF5DF7CFEFEF7E2DCC00000283AF5DF7;C=487;L=152;`

erste Nachricht ohne preamble, zweite Nachricht komplett 
`MC;LL=-975;LH=984;SL=-487;SH=482;D=AF5DF7CFEFEF7E2DCC00000283AF5DF7CFEFEF7E2DCC;C=487;L=???`

erste und zweite Nachricht  komplett 
`MC;LL=-975;LH=984;SL=-487;SH=482;D=000283AF5DF7CFEFEF7E2DCC00000283AF5DF7CFEFEF7E2DCC;C=487;L=???`


> Die Annahme, dass index immer schneller als eine regexp ist kann ich übrigens nicht bestätigen, ganz im Gegenteil ist es in diesem Fall doch erheblich langsamer sogar

Daß index schneller ist, habe ich mal irgendwo gelesen, aber anscheinend gibt es da auch Ausnahmen.

Lässt sich anhand der Rate deuten, daß das regex ca 2-3 mal schneller ist?
index_str3  2513115/s     
index_str1  2513115/s      
index_str2  3026401/s         
regexp_str3 6255709/s      
regexp_str1 6313100/s  
regexp_str2 6340946/s

Reply:
> Wenn man 25 oder mehr 1er Bits als Preamble empfängt, dann ist es nicht mehr das OSV3 Protokoll, dann ist es einfach etwas anderes.

Ist Dir ein solches Protokol bekannt, das eine Preamble mit mehr als 24 und den selben Sync hat.
Mir ist kein solches bekannt. 

Damit `($bitData =~ m/^.?(1){12,24}0101/)` funktioniert aber der o.g. Fall 3 nicht.

Wünsche frohe Rest Weihnachten

Reply:
OSV2 hat beispielsweise mehr als 24 Sync bits. Wie dem auch sei, was gewinnen wir denn außer ein alse positive zu riskieren. Die  Anpassung der Regex müsste dann so lauten:
`($bitData =~ m/1{12,24}(0101)/)` deckt den gewünschten Fall doch sauber ab.

```
$preamble_pos = $-[1];; 
```

Die Index Funktion wäre in diesem Fall dann tatsächlich schneller, aber auch wieder ungenauer.

Die Regex hier anzupassen und nicht nur vom Start weg zu suchen sollten wir genau beobachten. Es hat bestimmt einen Grund, weshalb ich das auf den Start der Bitdaten begrenzt hatte.

Was mich außerdem wundert ist, dass nicht beide Nachrichten komplett im Puffer sind. Ich habe nachgesehen, es passen 400 Manchester Bits in den Speicher. Also genügend für zwei Nachrichten.

Reply:
> OSV2 hat beispielsweise mehr als 24 Sync bits

OSV2 hat aber keine preamble mit einsen.

Ja, das `1{12,24}` dürfte nicht schaden, aber ich sehe auch keinen Nutzen.

> Es hat bestimmt einen Grund, weshalb ich das auf den Start der Bitdaten begrenzt hatte.

Der Grund dürfte gewesen sein, daß es damals die neuen OSV3 Sensoren noch nicht gab.
Seither hatten die OSV3 Sensoren keine Wiederholungen ohne Pause

Reply:
Bei OSV2 habe ich es aber auch so gemacht und da sind diese Sensoren bereits bekannt

`	if ($bitData =~ m/^.?(01){12,17}.?10011001/) `